bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadAheadManager.java

```diff
@@ -0,0 +1,655 @@
+        // check if we need to read ahead
+        boolean isHit = hitInReadAheadPositions(ledgerId, entryId);
+        if (log.isDebugEnabled()) {
+            if (isHit) {
+                log.debug("[L{} E{}] Trigger read-ahead task", ledgerId, entryId);
+            } else {
+                log.debug("[L{} E{}] Not found in pending map", ledgerId, entryId);
+            }
+        }
+
+        // hit in the cache
+        ByteBuf entry = cache.get(ledgerId, entryId);
+        if (entry != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("[L{} E{}] Hit cache directly", ledgerId, entryId);
+            }
+            dbLedgerStorageStats.getReadCacheHitCounter().inc();
+            return entry;
+        }
+        dbLedgerStorageStats.getReadCacheMissCounter().inc();
+
+        // search entry in the read-ahead index
+        ReadAheadTaskStatus readAheadTaskStatus = getNearestTask(ledgerId, entryId);
+        if (readAheadTaskStatus != null && readAheadTaskStatus.hasEntry(entryId)) {
+            if (log.isDebugEnabled()) {
+                log.debug("[L{} E{}] Block until data is ready", ledgerId, entryId);
+            }
+            readAheadTaskStatus.waitUntilReadCompleted(dbLedgerStorageStats);
+
+            entry = cache.get(ledgerId, entryId);
+            if (entry != null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[L{} E{}] Hit cache after read-ahead", ledgerId, entryId);
+                }
+                return entry;
+            }
+        }
+
+        // read the current pos entry, and add the next pos into read-ahead set
+        if (log.isDebugEnabled()) {
+            log.debug("[L{} E{}] Read from storage layer", ledgerId, entryId);
+        }
+        return readAndAddNextReadAheadPosition(ledgerId, entryId);
+    }
+
+    private void internalReadAhead(long ledgerId, long entryId, long location,
+                                   int maxMessages, int maxBytes, long submitStartNanos) {
+        // record queue time
+        recordStatsInNano(dbLedgerStorageStats.getReadAheadAsyncQueueTime(), submitStartNanos);
+        long readAheadStartNanos = MathUtils.nowInNano();
+
+        // obtain the read-ahead info
+        ReadAheadTaskStatus readAheadTaskStatus = getNearestTask(ledgerId, entryId);
+        if (readAheadTaskStatus == null) {
+            return;
+        }
+
+        // record all the positions of these entries
+        List<Long> entriesPos = new ArrayList<>();
+
+        // read from fc
+        int messages = 0;
+        int bytes = 0;
+
+        // flag to determine whether keeps reading ahead
+        boolean earlyExit = false;
+
+        // flag to indicate whether reads from cache since last time
+        boolean readFromCacheBefore = false;
+
+        try {
+            while (messages <= maxMessages && bytes <= maxBytes) {
+
+                ByteBuf entry = cache.get(ledgerId, entryId);
```

Comment:
We'd better use `cache.hasEntry` instead of `cache.get` to avoid unnecessary direct memory allocate.